sycl/include/CL/sycl/INTEL/ac_int.hpp

```diff
@@ -0,0 +1,2458 @@
+  return ap_equal_zeros_to<B, N>(v);
+}
+
+// if bits [0, B-1] all 0s
+template <int B, int N> constexpr bool ap_equal_ones_to(ap_uint<N> value) {
+  static_assert(N >= B, "");
+  if (!B)
+    return true;
+  constexpr int B1 = AC_MAX(B, 1);
+  ap_uint<B1> v = bit_slc<B1, 0, N>(value);
+  return (~v) == static_cast<ac_private::ap_uint<B1>>(0);
+}
+
+template <int B, int N> constexpr bool ap_equal_ones_to(ap_int<N> value) {
+  ap_uint<N> v = value;
+  return ap_equal_ones_to<B, N>(v);
+}
+
+// if bits [B, N-1] are all ones
+template <int B, int N> constexpr bool ap_equal_zeros_from(ap_uint<N> value) {
+  constexpr int L = AC_MAX(N - B, 1);
+  ap_uint<L> v = bit_slc<L, B, N>(value);
+  return v == static_cast<ac_private::ap_uint<L>>(0);
+}
+
+template <int B, int N> constexpr bool ap_equal_zeros_from(ap_int<N> value) {
+  ap_uint<N> v = value;
+  return ap_equal_zeros_from<B, N>(v);
+}
+
+template <int B, int N> constexpr bool ap_equal_ones_from(ap_uint<N> value) {
+  constexpr int L = AC_MAX(N - B, 1);
+  ap_uint<L> v = bit_slc<L, B, N>(value);
+  return (~v) == static_cast<ac_private::ap_uint<L>>(0);
+}
+
+template <int B, int N> constexpr bool ap_equal_ones_from(ap_int<N> value) {
+  ap_uint<N> v = value;
+  return ap_equal_ones_from<B, N>(v);
+}
+
+// Build an ap_int from double d, value is d * 2 ^ N
+// Notice: in ref, it's d * 2 ^ (32 * N)
+template <int N, bool S>
+inline void ap_conv_from_fraction(double d, ap_int<N> &r, bool *qb, bool *rbits,
+                                  bool *o, int *io) {
+  bool b = d < 0;
+  double d2 = b ? -d : d;
+  double dfloor = mgc_floor(d2);
+  *o = dfloor != 0.0;
+  d2 = d2 - dfloor;
+  const int shift_amount = N + 1;
+  const int container_length = (shift_amount + 31) / 32 * 32 + 1;
+  ap_int<N + 1> tb;
+  ap_uint<container_length> k1;
+  const Ulong u64_1 = 1;
+  // for performance purpose, manually unroll the loop when shift_amount <= 64
+  if (shift_amount <= 32) {
+    d2 *= u64_1 << shift_amount;
+    k1 = (unsigned int)floor(d2);
+    tb = b ? ~k1 : k1;
+    d2 -= k1;
+  } else if (shift_amount <= 64) {
+    d2 *= u64_1 << 32;
+    unsigned int temp = (unsigned int)floor(d2);
+    k1 = temp;
+    d2 -= temp;
+    const int shift_next = AC_MAX(shift_amount - 32, 0);
+    d2 *= u64_1 << shift_next;
+    temp = (unsigned int)floor(d2);
+    k1 <<= shift_next;
+    k1 |= temp;
+    d2 -= temp;
+    tb = b ? ~k1 : k1;
+  } else {
+    k1 = static_cast<ac_private::ap_uint<container_length>>(0);
+    int to_shift = shift_amount;
+    unsigned int temp;
+    while (to_shift >= 32) {
+      to_shift -= 32;
+      d2 *= u64_1 << 32;
+      temp = (unsigned int)floor(d2);
+      k1 <<= 32;
+      k1 |= temp;
+      d2 -= temp;
+    }
+    const int shift_next = AC_MAX(to_shift % 32, 0);
+    d2 *= u64_1 << shift_next;
+    temp = (unsigned int)floor(d2);
+    k1 <<= shift_next;
+    k1 |= temp;
+    d2 -= temp;
+    tb = b ? ~k1 : k1;
+  }
+
+  r = tb;
+  d2 *= 2;
+  bool k = (int(d2)) != 0; // math
```

Comment:
Yeah, the math indeed :)
not very informative comment, I suggest to remove it

Reply:
Done. Thanks!